### ISSUE TITLE
ReferenceError: window is not defined

### ISSUE THREAD

**Issue by sleignnet** [ably/ably-js-nativescript#13]
I recently upgraded an app from Nativescript 6 to Nativescript 8.  In this app, I use this ably plugin to communicate between a webview and an external website.  Everything worked perfectly in v6 but after upgrading to Nativescript 8, I've been getting this error:

 ```
   System.err: An uncaught Exception occurred on "main" thread.
    System.err: Calling js method onPageStarted failed
    System.err: ReferenceError: window is not defined
    System.err:
    System.err: StackTrace:
    System.err: ./node_modules/ably/browser/static/ably-nativescript.js(file: node_modules\ably\browser\static\ably-nativescript.js:21:1)
    System.err:     at __webpack_require__(file: app\webpack\bootstrap:24:0)
    System.err:     at fn(file: app\webpack\runtime\hot module replacement:61:0)
    System.err:     at ./node_modules/ably-nativescript/ably.js(file: node_modules\ably-nativescript\ably.js:2:0)
    System.err:     at __webpack_require__(file: app\webpack\bootstrap:24:0)
    System.err:     at fn(file: app\webpack\runtime\hot module replacement:61:0)
    System.err:     at ./app/views/eitin/itin_show/eitin.js(file: app\app\views\eitin\itin_show\eitin.js:25:11)
    System.err:     at __webpack_require__(file: app\webpack\bootstrap:24:0)
    System.err:     at fn(file: app\webpack\runtime\hot module replacement:61:0)
    System.err:     at webpackContext(file: app\app|sync|\.(xml|js|s:187:0)
    System.err:     at loader(file: node_modules\@nativescript\core\globals\index.js:158:0)
    System.err:     at loadModule(file: node_modules\@nativescript\core\globals\index.js:202:0)
    System.err:     at createViewFromEntry(file: node_modules\@nativescript\core\ui\builder\index.js:25:0)
    System.err:     at navigate(file: node_modules\@nativescript\core\ui\frame\frame-common.js:180:21)
    System.err:     at exports.checkItinRedirection(file: app\app\views\calendars\fullcalendar\fullcalendar.js:152:0)
    System.err:     at _handleEvent(file: node_modules\@nativescript\core\data\observable\index.js:233:0)
    System.err:     at notify(file: node_modules\@nativescript\core\data\observable\index.js:216:0)
    System.err:     at _onLoadStarted(file: node_modules\@nativescript\core\ui\web-view\web-view-common.js:28:0)
    System.err:     at WebViewClientImpl.onPageStarted(file: node_modules\@nativescript\core\ui\web-view\index.android.js:30:0)
    System.err:     at com.tns.Runtime.callJSMethodNative(Native Method)
    System.err:     at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1302)
    System.err:     at com.tns.Runtime.callJSMethodImpl(Runtime.java:1188)
    System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:1175)
    System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:1153)
    System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:1149)
    System.err:     at com.tns.gen.android.webkit.WebViewClient_vendor_54051_28_WebViewClientImpl.onPageStarted(WebViewClient_vendor_54051_28_WebViewClientImpl.java:40)
    System.err:     at ic.c(PG:314)
    System.err:     at An.handleMessage(PG:20)
    System.err:     at android.os.Handler.dispatchMessage(Handler.java:107)
    System.err:     at android.os.Looper.loop(Looper.java:214)
    System.err:     at android.app.ActivityThread.main(ActivityThread.java:7356)
    System.err:     at java.lang.reflect.Method.invoke(Native Method)
    System.err:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
    System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

I looked at the ably-nativescript plugin code itself, and the area that's throwing above error is this:

```
    (function webpackUniversalModuleDefinition(root, factory) {
    	if(typeof exports === 'object' && typeof module === 'object')
    		module.exports = factory(require("nativescript-websockets"));
    	else if(typeof define === 'function' && define.amd)
    		define(["nativescript-websockets"], factory);
    	else if(typeof exports === 'object')
    		exports["Ably"] = factory(require("nativescript-websockets"));
    	else
    		root["Ably"] = factory(root["nativescript-websockets"]);
    })(window, function(__WEBPACK_EXTERNAL_MODULE__40__)
```

The second this plugin is referenced I get that error, every time.  If I don't include this plugin, the error goes away.

I know that the window object doesn't really even exist in NS so I'm not sure why it's in this code, or how this even worked in v6, unless something was done in the older version of NS's webpack, since they upgraded webpack after v6.

To that effect, my current webpack configuration is fairly simplistic, so maybe I'm missing something here?

```
    const webpack = require("@nativescript/webpack");
    
    module.exports = (env) => {
    	webpack.init(env);

    	webpack.Utils.addCopyRule('fonts/**')
    	webpack.Utils.addCopyRule('**/*.jpg')
    	webpack.Utils.addCopyRule('**/*.png')
    	webpack.Utils.addCopyRule('html_views/**')
    	    
    	return webpack.resolveConfig();
    };
```

**Comment by tqhsonoma** [comment]
Any update on this issue? I am having the same problem. Here is my package.json for a simple test project:

"scripts": {
    "ios": "ns run ios --no-hmr",
    "android": "ns run android --no-hmr --compileSdk 30"
  },
  "name": "menu",
  "main": "./src/main.ts",
  "version": "1.0.0",
  "private": true,
  "dependencies": {
    "@angular/animations": "~13.3.0",
    "@angular/common": "~13.3.0",
    "@angular/compiler": "~13.3.0",
    "@angular/core": "~13.3.0",
    "@angular/forms": "~13.3.0",
    "@angular/platform-browser": "~13.3.0",
    "@angular/platform-browser-dynamic": "~13.3.0",
    "@angular/router": "~13.3.0",
    "@nativescript/angular": "^13.0.0",
    "@nativescript/core": "~8.2.0",
    "@nativescript/theme": "~3.0.2",
    "ably-nativescript": "^1.1.0",
    "nativescript-angular": "^10.0.0",
    "rxjs": "~7.5.0",
    "zone.js": "~0.11.5"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~13.3.3",
    "@angular/cli": "~13.3.3",
    "@angular/compiler-cli": "~13.3.0",
    "@nativescript/android": "~8.2.2",
    "@nativescript/ios": "~8.2.3",
    "@nativescript/types": "~8.2.0",
    "@nativescript/webpack": "~5.0.6",
    "@ngtools/webpack": "~13.2.0",
    "typescript": "~4.5.5"
  }
}

**Comment by owenpearson** [comment]
Hi @tqhsonoma and @sleignnet I'm really sorry for the delay on getting back to you on this one. We'll have a look in to this and let you know when we have a fix, but for now a workaround should be to force the underlying ably-js version to 1.2.4 or below. If you're using NPM 8.3+ you should be able to do this by adding the following to your package.json:
```json
{
  "overrides": {
    "ably": "1.2.4"
  }
}
```
If you're using an older version of NPM instructions will be slightly different, there are some solutions in [this stack overflow thread](https://stackoverflow.com/questions/15806152/how-do-i-override-nested-npm-dependency-versions). Let us know if you have any further issues.

**Comment by tqhsonoma** [comment]
Hi,

I am on NPM 8.7.0. Add the above lines as recommended to the package.json, did a "ns clean" and then ran "ns run ios". Different error this time. Any other suggestions?

ERROR in ./node_modules/ably/browser/static/ably-nativescript.js 3414:20-51
Module not found: Error: Can't resolve 'application-settings' in '/Users/thai/Projects/ns/menu/node_modules/ably/browser/static'
resolve 'application-settings' in '/Users/thai/Projects/ns/menu/node_modules/ably/browser/static'
  Parsed request is a module
  using description file: /Users/thai/Projects/ns/menu/node_modules/ably/package.json (relative path: ./browser/static)
    resolve as module
      looking for modules in /Users/thai/Projects/ns/menu/node_modules
        single file module
          using description file: /Users/thai/Projects/ns/menu/package.json (relative path: ./node_modules/application-settings)
            no extension
              /Users/thai/Projects/ns/menu/node_modules/application-settings doesn't exist
            .ios.ts
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.ts doesn't exist
            .ts
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ts doesn't exist
            .ios.js
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.js doesn't exist
            .js
              /Users/thai/Projects/ns/menu/node_modules/application-settings.js doesn't exist
            .ios.mjs
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.mjs doesn't exist
            .mjs
              /Users/thai/Projects/ns/menu/node_modules/application-settings.mjs doesn't exist
            .ios.css
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.css doesn't exist
            .css
              /Users/thai/Projects/ns/menu/node_modules/application-settings.css doesn't exist
            .ios.scss
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.scss doesn't exist
            .scss
              /Users/thai/Projects/ns/menu/node_modules/application-settings.scss doesn't exist
            .ios.json
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.json doesn't exist
            .json
              /Users/thai/Projects/ns/menu/node_modules/application-settings.json doesn't exist
        /Users/thai/Projects/ns/menu/node_modules/application-settings doesn't exist
      /Users/thai/Projects/ns/menu/node_modules/ably/browser/static/node_modules doesn't exist or is not a directory
      /Users/thai/Projects/ns/menu/node_modules/ably/browser/node_modules doesn't exist or is not a directory
      /Users/thai/Projects/ns/menu/node_modules/ably/node_modules doesn't exist or is not a directory
      /Users/thai/Projects/ns/menu/node_modules/node_modules doesn't exist or is not a directory
      looking for modules in /Users/thai/Projects/ns/menu/node_modules
        single file module
          using description file: /Users/thai/Projects/ns/menu/package.json (relative path: ./node_modules/application-settings)
            no extension
              /Users/thai/Projects/ns/menu/node_modules/application-settings doesn't exist
            .ios.ts
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.ts doesn't exist
            .ts
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ts doesn't exist
            .ios.js
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.js doesn't exist
            .js
              /Users/thai/Projects/ns/menu/node_modules/application-settings.js doesn't exist
            .ios.mjs
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.mjs doesn't exist
            .mjs
              /Users/thai/Projects/ns/menu/node_modules/application-settings.mjs doesn't exist
            .ios.css
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.css doesn't exist
            .css
              /Users/thai/Projects/ns/menu/node_modules/application-settings.css doesn't exist
            .ios.scss
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.scss doesn't exist
            .scss
              /Users/thai/Projects/ns/menu/node_modules/application-settings.scss doesn't exist
            .ios.json
              /Users/thai/Projects/ns/menu/node_modules/application-settings.ios.json doesn't exist
            .json
              /Users/thai/Projects/ns/menu/node_modules/application-settings.json doesn't exist
        /Users/thai/Projects/ns/menu/node_modules/application-settings doesn't exist
      /Users/thai/Projects/ns/node_modules doesn't exist or is not a directory
      /Users/thai/Projects/node_modules doesn't exist or is not a directory
      looking for modules in /Users/thai/node_modules
        single file module
          using description file: /Users/thai/package.json (relative path: ./node_modules/application-settings)
            no extension
              /Users/thai/node_modules/application-settings doesn't exist
            .ios.ts
              /Users/thai/node_modules/application-settings.ios.ts doesn't exist
            .ts
              /Users/thai/node_modules/application-settings.ts doesn't exist
            .ios.js
              /Users/thai/node_modules/application-settings.ios.js doesn't exist
            .js
              /Users/thai/node_modules/application-settings.js doesn't exist
            .ios.mjs
              /Users/thai/node_modules/application-settings.ios.mjs doesn't exist
            .mjs
              /Users/thai/node_modules/application-settings.mjs doesn't exist
            .ios.css
              /Users/thai/node_modules/application-settings.ios.css doesn't exist
            .css
              /Users/thai/node_modules/application-settings.css doesn't exist
            .ios.scss
              /Users/thai/node_modules/application-settings.ios.scss doesn't exist
            .scss
              /Users/thai/node_modules/application-settings.scss doesn't exist
            .ios.json
              /Users/thai/node_modules/application-settings.ios.json doesn't exist
            .json
              /Users/thai/node_modules/application-settings.json doesn't exist
        /Users/thai/node_modules/application-settings doesn't exist
      /Users/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory
 @ ./src/app/item/items.component.ts 2:0-65 31:27-40
 @ ./src/app/app-routing.module.ts 2:0-56 7:32-46
 @ ./src/app/app.module.ts 2:0-56 9:99-115
 @ ./src/main.ts 4:0-45 6:69-78

webpack 5.70.0 compiled with 1 error in 28672 ms
Webpack compilation complete. Watching for file changes.

**Comment by tqhsonoma** [comment]
Thank you, @owenpearson

**Comment by owenpearson** [comment]
@tqhsonoma @sleignnet Hi both, I've just released `ably-nativescript` version 1.2.0 which should resolve the above issues. Please upgrade to the new version and let us know if you have any further issues!

**Comment by tqhsonoma** [comment]
Hi @owenpearson,

Thank you for the new release. I think you will also need to update the imports for the application-settings. I had to manually make the following changes to get the updated Ably 1.2 plugin to work with the most current version of Nativescript 8.2:

node_modules/ably/browser/static/ably-nativescript.js
line #3482:
From: var appSettings = require("application-settings");
To: var appSettings = require("@nativescript/core/application-settings");


node_modules/ably/browser/lib/util/nativescript-webstorage.js
line #2: 
From: var appSettings = require("application-settings");
To: 	var appSettings = require("@nativescript/core/application-settings");

**Comment by owenpearson** [comment]
@tqhsonoma OK, thanks for letting us know. Funnily enough, I noted this as a potential issue but the old import seemed to work fine for me locally on the latest version of nativescript so I assumed it was okay 🤷  I've opened [ably-js#979](https://github.com/ably/ably-js/issues/979) to track this and should be able to release with the updated import within the next week. Let us know if you don't have a workaround and if this is high priority for you; if so I'll try and prioritise.

**Comment by owenpearson** [comment]
Hey @tqhsonoma, we've released ably-js-nativescript 1.2.1 which uses the new import path. Please get in touch if you have any further issues 🙂
